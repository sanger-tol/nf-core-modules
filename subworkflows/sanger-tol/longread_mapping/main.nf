include { CRAMALIGN_GENCRAMCHUNKS         } from '../../../modules/sanger-tol/cramalign/gencramchunks'
include { CRAMALIGN_MINIMAP2ALIGN         } from '../../../modules/sanger-tol/cramalign/minimap2align/main'
include { MINIMAP2_INDEX                  } from '../../../modules/nf-core/minimap2/index/main'
include { SAMTOOLS_FAIDX                  } from '../../../modules/nf-core/samtools/faidx/main'
include { SAMTOOLS_INDEX                  } from '../../../modules/nf-core/samtools/index/main'
include { SAMTOOLS_MERGE                  } from '../../../modules/nf-core/samtools/merge/main'

workflow LONGREAD_MAPPING {

    take:
    ch_assemblies           // Channel [meta, assembly]
    ch_crams                // Channel [meta, cram] OR [meta, [cram1, cram2, ..., cram_n]]
    val_cram_chunk_size     // integer: Number of CRAM slices per chunk for mapping

    main:
    ch_versions = Channel.empty()

    //
    // Logic: check if CRAM files are accompanied by an index
    //        Get indexes, and index those that aren't
    //
    ch_cram_raw = ch_crams
        | transpose()
        | branch { meta, cram ->
            def cram_file = file(cram, checkIfExists: true)
            def index = cram + ".crai"
            have_index: file(index).exists()
                return [ meta, cram_file, file(index, checkIfExists: true) ]
            no_index: true
                return [ meta, cram_file ]
        }

    //
    // Module: Index CRAM files without indexes
    //
    SAMTOOLS_INDEX(ch_cram_raw.no_index)
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions)

    ch_cram_indexed = ch_cram_raw.have_index
        | mix(
            ch_cram_raw.no_index.join(SAMTOOLS_INDEX.out.crai)
        )

    //
    // Module: Process the cram index files to determine how many
    //         chunks to split into for mapping
    //
    ch_cram_indexes = ch_cram_indexed
        | map { meta, cram, index -> [ meta, index ] }

    CRAMALIGN_GENCRAMCHUNKS(
        ch_cram_indexes,
        val_cram_chunk_size
    )
    ch_versions = ch_versions.mix(CRAMALIGN_GENCRAMCHUNKS.out.versions)

    //
    // Logic: Count the total number of cram chunks for downstream grouping
    //
    ch_n_cram_chunks = CRAMALIGN_GENCRAMCHUNKS.out.cram_slices
        | map { meta, chunkn, _slices -> [ meta, chunkn ] }
        | transpose()
        | groupTuple(by: 0)
        | map { meta, chunkns -> [ meta, chunkns.size() ] }

    //
    // Logic: Re-join the cram files and indexes to their chunk information
    //
    ch_cram_with_slices = ch_cram_indexed
        | combine(CRAMALIGN_GENCRAMCHUNKS.out.cram_slices, by: 0)

    //
    // MODULE: generate minimap2 mmi file
    //
    MINIMAP2_INDEX(ch_assemblies)
    ch_versions = ch_versions.mix(MINIMAP2_INDEX.out.versions)

    ch_cram_chunks = ch_cram_with_slices
        | transpose()
        | combine(MINIMAP2_INDEX.out.index, by: 0)

    CRAMALIGN_MINIMAP2ALIGN(ch_cram_chunks)
    ch_versions = ch_versions.mix(CRAMALIGN_MINIMAP2ALIGN.out.versions)

    //
    // Module: Index assembly fastas
    //
    SAMTOOLS_FAIDX(
        ch_assemblies, // reference
        [ [:],[] ],    // fai
        false          // get sizes
    )
    ch_versions = ch_versions.mix(SAMTOOLS_FAIDX.out.versions)

    //
    // Logic: create a channel with both fai and gzi for each assembly
    //        We do it here so we don't cause downstream issues with the
    //        remainder join
    //
    ch_fai_gzi = SAMTOOLS_FAIDX.out.fai
        | join(SAMTOOLS_FAIDX.out.gzi, by: 0, remainder: true)
        | map { meta, fai, gzi -> [ meta, fai, gzi ?: [] ] }


    //
    // Logic: Prepare input for merging bams.
    //        We use the ch_n_cram_chunks to set a groupKey so that
    //        we emit groups downstream ASAP once all bams have been made
    //
    ch_samtools_merge_input = CRAMALIGN_MINIMAP2ALIGN.out.bam
        | combine(ch_n_cram_chunks, by: 0)
        | map { meta, bam, n_chunks ->
            def key = groupKey(meta, n_chunks)
            [key, bam]
        }
        | groupTuple(by: 0)
        | map { key, bam -> [key.target, bam] } // Get meta back out of groupKey
        | combine(ch_assemblies, by: 0)
        | combine(ch_fai_gzi, by: 0)
        | multiMap { meta, bams, assembly, fai, gzi ->
            bam:   [ meta, bams ]
            fasta: [ meta, assembly ]
            fai:   [ meta, fai ]
            gzi:   [ meta, gzi ]
        }

    //
    // Module: Merge position-sorted bam files
    //
    SAMTOOLS_MERGE(
        ch_samtools_merge_input.bam,
        ch_samtools_merge_input.fasta,
        ch_samtools_merge_input.fai,
        ch_samtools_merge_input.gzi,
    )
    ch_versions = ch_versions.mix(SAMTOOLS_MERGE.out.versions)

    emit:
    bam      = SAMTOOLS_MERGE.out.bam
    versions = ch_versions
}
