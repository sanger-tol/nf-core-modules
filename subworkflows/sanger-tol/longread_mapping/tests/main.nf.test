nextflow_workflow {

    name "Test Subworkflow LONGREAD_MAPPING"
    script "../main.nf"
    workflow "LONGREAD_MAPPING"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_sangertol"
    tag "subworkflows/longread_mapping"
    tag "cramalign/gencramchunks"
    tag "cramalign/minimap2alignlongread"
    tag "subworkflows/../../modules/nf-core/minimap2/index"
    tag "subworkflows/../../modules/nf-core/samtools/merge"
    tag "subworkflows/../../modules/nf-core/samtools/index"
    tag "subworkflows/../../modules/nf-core/samtools/faidx"

    setup {
        nfcoreInitialise("${launchDir}/library/")
        nfcoreInstall(
            "${launchDir}/library/",
            [
                "minimap2/index",
                "samtools/index",
                "samtools/merge",
                "samtools/faidx"
            ]
        )
        nfcoreLink("${launchDir}/library/", "${baseDir}/modules/")
    }

    test("meles meles - single test") {

        when {
            params {
                samtools_cat_args     = ""
                samtools_fastq_args   = "-F 0x200 -nt"
                minimap2_args         = "-ax map-ont"
                samtools_sort_args    = "--write-index -l1"
            }

            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/ont/PAE35587_pass_1f1f0707_115.subset.cram', checkIfExists: true)
                    ]
                )
                input[2] = 10
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bam.size() == 1 },
                {
                    assert snapshot(
                        workflow.out.bam.collect { bam(it.get(1)).getStatistics() }.sort(),
                        workflow.out.versions
                    ).match()
                }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

    test("meles meles - multiple test") {

        when {
            params {
                samtools_cat_args     = ""
                samtools_fastq_args   = "-F 0x200 -nt"
                minimap2_args         = "-ax map-ont"
                samtools_sort_args    = "--write-index -l1"
            }

            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ],
                    [
                        [ id:'test1' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/ont/PAE35587_pass_1f1f0707_115.subset.cram', checkIfExists: true)
                    ],
                    [
                        [ id:'test1' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/ont/PAE35587_pass_1f1f0707_115.subset.cram', checkIfExists: true)
                    ]
                )
                input[2] = 5
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bam.size() == 2 },
                {
                    assert snapshot(
                        workflow.out.bam.collect { bam(it.get(1)).getStatistics() }.sort(),
                        workflow.out.versions
                    ).match()
                }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

    test("meles meles - mismatch meta test") {

        when {
            params {
                samtools_cat_args     = ""
                samtools_fastq_args   = "-F 0x200 -nt"
                minimap2_args         = "-ax map-ont"
                samtools_sort_args    = "--write-index -l1"
            }

            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ],
                    [
                        [ id:'test1' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/ont/PAE35587_pass_1f1f0707_115.subset.cram', checkIfExists: true)
                    ],
                    [
                        [ id:'test2' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/ont/PAE35587_pass_1f1f0707_115.subset.cram', checkIfExists: true)
                    ]
                )
                input[2] = 5
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bam.size() == 1 },
                {
                    assert snapshot(
                        workflow.out.bam.collect { bam(it.get(1)).getStatistics() }.sort(),
                        workflow.out.versions
                    ).match()
                }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

    test("meles meles - PacBio Hifi test") {

        when {
            params {
                samtools_cat_args     = ""
                samtools_fastq_args   = "-F 0x200 -nt"
                minimap2_args         = "-ax map-hifi"
                samtools_sort_args    = "--write-index -l1"
            }

            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ],
                    [
                        [ id:'test1' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/pacbio/m64094_200911_174739.ccs.bc1022_BAK8B_OA--bc1022_BAK8B_OA.subset.cram', checkIfExists: true)
                    ],
                    [
                        [ id:'test1' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/pacbio/m64094_200911_174739.ccs.bc1022_BAK8B_OA--bc1022_BAK8B_OA.subset.cram', checkIfExists: true)
                    ]
                )
                input[2] = 5
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bam.size() == 2 },
                {
                    assert snapshot(
                        workflow.out.bam.collect { bam(it.get(1)).getStatistics() }.sort(),
                        workflow.out.versions
                    ).match()
                }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }
    

}
