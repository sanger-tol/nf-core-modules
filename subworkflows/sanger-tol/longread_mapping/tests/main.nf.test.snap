{
    "meles meles - multiple test": {
        "content": [
            [
                {
                    "maxReadLength": 96980,
                    "minReadLength": 0,
                    "meanReadLength": 4840,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 3,
                    "readCount": 16265,
                    "duplicateReadCount": 0,
                    "sorted": false
                },
                {
                    "maxReadLength": 96980,
                    "minReadLength": 0,
                    "meanReadLength": 4840,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 3,
                    "readCount": 16265,
                    "duplicateReadCount": 0,
                    "sorted": false
                }
            ],
            [
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,1e0ef2d568832aca8d333501199ae77b",
                "versions.yml:md5,1e0ef2d568832aca8d333501199ae77b",
                "versions.yml:md5,663fb1b828914274912d0ca51115378a",
                "versions.yml:md5,663fb1b828914274912d0ca51115378a",
                "versions.yml:md5,67afe4aba4aab8e07765f229e5964159",
                "versions.yml:md5,67afe4aba4aab8e07765f229e5964159",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-07T16:51:48.531568"
    },
    "meles meles - PacBio Hifi test": {
        "content": [
            [
                {
                    "maxReadLength": 32471,
                    "minReadLength": 0,
                    "meanReadLength": 5954,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 3,
                    "readCount": 26719,
                    "duplicateReadCount": 0,
                    "sorted": false
                },
                {
                    "maxReadLength": 32471,
                    "minReadLength": 0,
                    "meanReadLength": 5954,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 3,
                    "readCount": 26719,
                    "duplicateReadCount": 0,
                    "sorted": false
                }
            ],
            [
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,1e0ef2d568832aca8d333501199ae77b",
                "versions.yml:md5,1e0ef2d568832aca8d333501199ae77b",
                "versions.yml:md5,663fb1b828914274912d0ca51115378a",
                "versions.yml:md5,663fb1b828914274912d0ca51115378a",
                "versions.yml:md5,67afe4aba4aab8e07765f229e5964159",
                "versions.yml:md5,67afe4aba4aab8e07765f229e5964159",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-07T16:55:47.124802"
    },
    "meles meles - mismatch meta test": {
        "content": [
            [
                {
                    "maxReadLength": 96980,
                    "minReadLength": 0,
                    "meanReadLength": 4840,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 3,
                    "readCount": 16265,
                    "duplicateReadCount": 0,
                    "sorted": false
                }
            ],
            [
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,1e0ef2d568832aca8d333501199ae77b",
                "versions.yml:md5,1e0ef2d568832aca8d333501199ae77b",
                "versions.yml:md5,663fb1b828914274912d0ca51115378a",
                "versions.yml:md5,663fb1b828914274912d0ca51115378a",
                "versions.yml:md5,67afe4aba4aab8e07765f229e5964159",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-07T16:52:48.5016"
    },
    "meles meles - single test": {
        "content": [
            [
                {
                    "maxReadLength": 96980,
                    "minReadLength": 0,
                    "meanReadLength": 4840,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 3,
                    "readCount": 16265,
                    "duplicateReadCount": 0,
                    "sorted": false
                }
            ],
            [
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,1e0ef2d568832aca8d333501199ae77b",
                "versions.yml:md5,663fb1b828914274912d0ca51115378a",
                "versions.yml:md5,67afe4aba4aab8e07765f229e5964159",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd",
                "versions.yml:md5,bb93856f33590e258c1f2c817bcf85cd"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-07T16:50:19.705145"
    }
}